README.md
# NAME

Cot - super lightweight perl framework based on Plack

# SYNOPSIS

    #!/usr/bin/env perl -w

    use Cot;

    get '/' => sub {
        my $self = shift;
        $self->res->status(200);
        $self->res->headers( { 'Content-Type' => 'text/plain', } );
        $self->res->body('Hello world!');
    };

    run;

The above is a basic but functional web app created with Cot.

# DESCRIPTION

Cot is super lightweight perl framework base on Plack!
you can extend Cot by using many plugins.





You can initialize application skelton using cotto utility executable.

    % cot init Test
    % cd ./Test
    % cot run

# METHODS

## plugin

Added plugins to Cot App.

    use Cot;

    plugin qw/Config Session/;

## get

Receive GET request:

    use Cot;

    get '/api/echo' => sub {
        my $self = shift;
        #code
    };

## post

Receive POST request:

    use Cot;

    post '/api/echo' => sub {
        my $self = shift;
        #code
    };

## put

Receive PUT request:

    use Cot;

    put '/api/echo' => sub {
        my $self = shift;
        #code
    };

## delete

Receive DELETE request:

    use Cot;

    delete '/api/echo' => sub {
        my $self = shift;
        #code
    };

## options

Receive OPTIONS request:

    use Cot;

    options '/api/echo' => sub {
        my $self = shift;
        #code
    };

## patch

Receive PATCH request:

    use Cot;

    patch '/api/echo' => sub {
        my $self = shift;
        #code
    };

## any

Receive any HTTP request:

    use Cot;

    any '/api/echo' => sub {
        my $self = shift;
        #code
    };

## static

Serve static files:

    use Cot;

    static => '/ui';

## run

Up the Plack execution loop.
You can set [plackup](http://search.cpan.org/perldoc?plackup) arguments.

    use Cot;

    run("-- port 5001 -R public");

## forbidden\_response

    use Cot;

    get '/secret' => sub {
        $self->res->status(403);
        $self->res->body('forbidden');
    };

same as below.

    use Cot;

    get '/secret' => sub {
        $self->forbidden_response;
    };

## notfound\_response

    use Cot;

    get '/secret' => sub {
        $self->res->status(404);
        $self->res->body('notfound');
    };

same as below.

    use Cot;

    get '/secret' => sub {
        $self->notfound_response;
    };

## redirect\_response

    use Cot;

    get '/secret' => sub {
        $self->res->redirect('/', 301);
    };

same as below.

    use Cot;

    get '/secret' => sub {
        $self->redirect_response;
    };

# Context METHODS

## req

__req__ is a [Plack::Request](http://search.cpan.org/perldoc?Plack::Request) Object. You can call all methods of Plack::Request.

    use Cot;

    get '/test' => sub {
        my $self = shift;
        my $req = $self->req;
        my $test = $req->param('test');
    };

## res

__res__ is a [Plack::Response](http://search.cpan.org/perldoc?Plack::Response) Object. You can call all methods of Plack::Response.

    use Cot;

    get '/test' => sub {
        my $self = shift;
        $self->res->status(200);
        $self->res->headers({'Content-Type' => 'text/plain' });
        $self->res->body($self->config->{sample});
    };

## path\_info

__path\_info__ is PATH\_INFO Array ref object.

    use Cot;

    # if called /test/hello
    get '/test' => sub {
        my $self = shift;
        my $info = $self->path_info->[0]; #hello
        ...
    };

## env

__env__ is Plack environment variable.

    use Cot;

    get '/test' => sub {
        my $self = shift;
        my $remote_addr = $self->env->{REMOTE_ADDR}; # same as $self->req->address
        ...
    };

## path

__path__ is requested PATH string

    use Cot;

    # if called /test/hello/myname
    get '/test' => sub {
        my $self = shift;
        my $path = $self->path; # /test/hello/myname
        ...
    };

# ENV

You can set ENVIRONMENT variables for change behaviour.

## COT\_ROOT

Default value is " B <.>". For example mod\_perl configuration, you can set

    PerlSetEnv COT_ROOT /www/TestApp/

## COT\_ENV

Default value is __developement__. You can change COT\_ENV for configration.

    #!/bin/sh
    export COT_ENV=production
    cot run

## COT\_DIRECTORYINDEX

Default value is none. If you use __static__ method, automatically serve DIRECTORYINDEX

    #!/bin/sh
    export DIRECTORYINDEX=index.html:index.xhtml
    cot run

# PLUGINS

Cot has plaggable interface. For default install only [Cot::Plugin::Config](http://search.cpan.org/perldoc?Cot::Plugin::Config) can be used.

config file($ENV{COT\_ROOT}/config.yaml):

    developement:
      hello: world

application code:

    use Cot;
    use Cot::Plugin qw/Config/;
    # or use Cot::Plugin::Config;

    get '/' => sub {
       my $self = shift;
       my $hello = $c->config->{'hello'}; # world
       ...
    };

# AUTHORS

This module has been written by Yusuke Shibata <shibata@yusukeshibata.jp> and others,
see the AUTHORS file that comes with this distribution for details.

# SOURCE CODE

The source code for this module is hosted on GitHub
[https://github.com/yusukeshibata/Cot](https://github.com/yusukeshibata/Cot).  Feel free to fork the repository and
submit pull requests!

# DEPENDENCIES

The following modules are mandatory (Cot cannot run without them):

- [YAML](http://search.cpan.org/perldoc?YAML)
- [Plack](http://search.cpan.org/perldoc?Plack)

# LICENSE

Copyright (C) Yusuke Shibata

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Yusuke Shibata <shibata@yusukeshibata.jp>


Changes
Revision history for Perl extension Cot

0.11 2013-09-26T02:44:05Z

   - indexes method

0.10 2013-08-12T13:36:00Z

   - for cgi mode.

0.09 2013-07-22T10:41:19Z

   - generated up script updated.

0.08 2013-04-26T00:51:02Z

   - changed script and document updated.

0.07 2013-04-26T00:47:44Z

   - plugin method added.

0.06 2013-04-25T02:19:30Z

   - "uri" method renamed "path"
   - [redirect|notfound|forbidden]_response added.
   - COT_DIRECTORYINDEX bug fixed.

0.05 2013-04-24T00:54:40Z

   - test updated.

0.04 2013-04-23T06:18:24Z

   - cot script generate 'up' refined.
   - PUT,DELETE,PATCH,OPTIONS method accept.

0.03 2013-04-23T02:03:30Z

   - cot script updated.

0.02 2013-04-22T12:39:36Z

    - fix document typos.
    - plugin loader implemented.

0.01 2013-04-22T05:11:27Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Cot',
    module_name     => 'Cot',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Cot/Plugin.pm
package Cot::Plugin;

use strict;
use warnings;
use 5.008005;
our $VERSION = "0.11";
$VERSION = eval $VERSION;
use Carp;

sub import {
    my $class = shift;
    my $pkg   = caller(0);
    if ( $class eq 'Cot::Plugin' ) {
        foreach (@_) {
            my $klass = "Cot::Plugin::$_";
            eval "require $klass" or croak "Plugin[$_] is not installed.";
            $klass->_regist($pkg);
        }
        return;
    }
    $class->_regist($pkg);
}

sub _regist {
    my ( $class, $pkg ) = @_;
    croak "$pkg is not Cot object[$class importing...]"
      unless $pkg->isa('Cot');
    $pkg->_register_plugin($class);
}
sub new { bless {}, shift; }

sub init {
    my ( $self, $c ) = @_;
}

1;
__END__

=encoding utf-8

=head1 NAME

Cot::Plugin - Cot plugin base module

=head1 SYNOPSIS

For example, Text::Xslate plugin is below(It's only a example!).

    use strict;
    use warnings;
    use parent Cot::Plugin;
    use Text::Xslate;

    ### "init" is initialize method only called once.

    sub init {
         my ( $self, $c ) = @_;
         # in "init" adding "tx" method to Cot context object.
         $c->tx($self);
    }

    ### define plugin methods

    # "tx" method can call "output";
    sub output {
        my $self          = shift;
        my $txfile        = shift;
        my $param         = shift;
        my $tx            = Text::Xslate->new;
        $tx->render( $txfile, $param );
     }
     1;
     __END__

You can use in a Cot application.

    use Cot;
    use Cot::Plugin qw/TX/;
    # or use Cot::Plugin::TX;

    get '/' => sub {
        my $self = shift;
        $self->res->status(200);
        $self->res->headers({'Content-Type' => 'text/plain' });

        # you can call "tx" plugin
        my $out = $self->tx->output('/path/to/template', { hello => 'world' });
        $self->res->body($out);
    };

=head1 DESCRIPTION

Cot::Plugin is base class of Plugins.

=head1 LICENSE

Copyright (C) Yusuke Shibata

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Yusuke Shibata E<lt>shibata@yusukeshibata.jpE<gt>



lib/Cot/Plugin/Config.pm
package Cot::Plugin::Config;

use strict;
use warnings;
use 5.008005;
our $VERSION = "0.11";
$VERSION = eval $VERSION;
use parent qw(Cot::Plugin);
use Cot::Config;

sub init {
    my ( $self, $c ) = @_;
    $c->config( Cot::Config->config );
}

1;
__END__

=encoding utf-8

=head1 NAME

Cot::Plugin::Config - Cot configuration plugin

=head1 SYNOPSIS

For example,

    use Cot;
    use Cot::Plugin qw/Config/;

    get "/test" => sub {
        my $self = shift;
        my $db = $self->config->{'db'};
        ...
    };

Config yaml file($ENV{COT_ROOT}/config.yaml) is,

    development:
      db: dbi:mysql:diary:localhost

    production:
      db: dbi:mysql:diary:192.168.1.10

=head1 DESCRIPTION

Cot::Plugin::Config is an only build in Plugin is Cot framework.
You can descsribe Config file(B<config.yaml>) and call B<config> method of Cot object.
return the hash reference of B<config.yaml> file $ENV{COT_ENV} entries data.

=head1 LICENSE

Copyright (C) Yusuke Shibata

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Yusuke Shibata E<lt>shibata@yusukeshibata.jpE<gt>



lib/Cot.pm
package Cot;

use strict;
use warnings;
use 5.008005;
our $VERSION = "0.11";
$VERSION = eval $VERSION;
use File::Spec;
use Plack::Request;
use Plack::Runner;
use Plack::App::File;
use DirHandle;
use Carp;
use vars qw($AUTOLOAD %POOL);

sub import {
    my $class = shift;
    my $pkg   = caller(0);
    strict->import;
    warnings->import;
    {
        no strict 'refs';
        push @{"$pkg\::ISA"}, $class;
    }
    for my $func (qw/run get post put patch delete options any static indexes plugin/) {
        no strict 'refs';
        *{"$pkg\::$func"} = \&{"_export_$func"};
    }
}

sub _register_plugin {
    my ( $class, $plugin_klass ) = @_;
    $plugin_klass->new->init( $class->_app );
}

sub _root {
    $ENV{COT_ROOT} || '.';
}

sub _app {
    my $class = shift;
    $POOL{ $class->_root } || $class->new;
}

sub new {
    my $class = shift;
    my $self  = bless {
        controller => {},
        plugins    => [],
    }, $class;
    $POOL{ $class->_root } = $self;
}

# get '/' => sub { my $c = shift; }
sub _method {
    my ( $class, $path, $sub, @methods ) = @_;
    foreach (@methods) {
        $class->_app->{controller}->{$_} ||= {};
        $class->_app->{controller}->{$_}->{$path} = $sub;
    }
}

sub _export_plugin {
    my @plugins = @_;
    my $class   = caller(0);
    foreach (@_) {
        my $klass = "Cot::Plugin::$_";
        eval "require $klass" or croak "Plugin[$_] is not installed.";
        $klass->_regist($class);
    }
}

sub _export_get {
    my ( $path, $sub ) = @_;
    my $class = caller(0);
    $class->_method( $path, $sub, 'get' );
}

sub _export_post {
    my ( $path, $sub ) = @_;
    my $class = caller(0);
    $class->_method( $path, $sub, 'post' );
}

sub _export_put {
    my ( $path, $sub ) = @_;
    my $class = caller(0);
    $class->_method( $path, $sub, 'put' );
}

sub _export_delete {
    my ( $path, $sub ) = @_;
    my $class = caller(0);
    $class->_method( $path, $sub, 'delete' );
}

sub _export_patch {
    my ( $path, $sub ) = @_;
    my $class = caller(0);
    $class->_method( $path, $sub, 'patch' );
}

sub _export_options {
    my ( $path, $sub ) = @_;
    my $class = caller(0);
    $class->_method( $path, $sub, 'options' );
}

sub _export_any {
    my ( $path, $sub ) = @_;
    my $class = caller(0);
    $class->_method( $path, $sub, 'get', 'post', 'put', 'patch', 'options',
        'delete' );
}

sub _export_static {
    my ($path)     = @_;
    my $class      = caller(0);
    my $controller = $class->_app->{controller};
    $controller->{get} ||= {};
    $controller->{get}->{$path} = \&_static;
}

sub _export_indexes {
    my ($path)     = @_;
    my $class      = caller(0);
    my $controller = $class->_app->{controller};
    $controller->{get} ||= {};
    $controller->{get}->{$path} = \&_indexes;
}
sub _indexes {
    my $self      = shift;
    my $path_info = $self->env->{PATH_INFO};
    my $path =
      File::Spec->catfile( $ENV{DOCUMENT_ROOT} || 'public', $path_info );
    if ( !-e $path ) {
        $self->notfound_response;
    }
    elsif ( -d $path ) {
        if ( $path_info =~ /.*\/$/ ) {
	  my $body = '<ul>';
	  my $dh = DirHandle->new($path);
	  while (defined($_ = $dh->read)) { $body .= sprintf('<li><a href="%s">%s</a></li>', $_,$_); }
	  $body .= '</ul>';
	  $self->res->status(200);
	  $self->res->headers({'Content-Type' => 'text/html', charset => 'utf-8'});
	  $self->res->body($body);
	  return;
        }
        else {
            $self->redirect_response( $path_info . '/' );
        }
    }
    else {
        my $file = Plack::App::File->new( file => $path )->call( $self->env );
        $self->res->status( $file->[0] );
        $self->res->headers( $file->[1] );
        $self->res->body( $file->[2] );
    }

}
sub _static {
    my $self      = shift;
    my $path_info = $self->env->{PATH_INFO};
    my $path =
      File::Spec->catfile( $ENV{DOCUMENT_ROOT} || 'public', $path_info );
    if ( !-e $path ) {
        $self->notfound_response;
    }
    elsif ( -d $path ) {
        if ( $path_info =~ /.*\/$/ ) {
            my @di = split( /:/, $ENV{COT_DIRECTORYINDEX} || '' );
            foreach my $di (@di) {
                my $index = File::Spec->catfile( $path, $di );
                if ( -f $index ) {
                    my $file =
                      Plack::App::File->new( file => $index )
                      ->call( $self->env );
                    $self->res->status( $file->[0] );
                    $self->res->headers( $file->[1] );
                    $self->res->body( $file->[2] );
                    return;
                }
            }
            $self->forbidden_response;
        }
        else {
            $self->redirect_response( $path_info . '/' );
        }
    }
    else {
        my $file = Plack::App::File->new( file => $path )->call( $self->env );
        $self->res->status( $file->[0] );
        $self->res->headers( $file->[1] );
        $self->res->body( $file->[2] );
    }

}

sub app {
    my ( $class, $env ) = @_;
    my $self      = $class->_app;
    my @path_info = ();
    my $req       = Plack::Request->new($env);
    my $method    = lc( $req->method );
    my $path      = $req->uri->path;
    $path =~ s/$env->{SCRIPT_NAME}// if $env->{SCRIPT_NAME};
    my @path = File::Spec->splitdir($path);
    my $controllers = $self->{controller}->{$method} || {};
    my $controller;

    for ( ; ; ) {
        my $u = File::Spec->catdir(@path);
        $controller = $controllers->{$u} and last;
        last unless scalar(@path);
        unshift @path_info, pop(@path);
    }
    $self->{req}       = $req;
    $self->{env}       = $env;
    $self->{res}       = $req->new_response;
    $self->{path}      = $path;
    $self->{path_info} = \@path_info;
    $controller ? &{ \&$controller }($self) : $self->forbidden_response;
    $self->res->finalize;
}
sub req       { shift->{req}; }
sub res       { shift->{res}; }
sub env       { shift->{env}; }
sub path      { shift->{path}; }
sub path_info { shift->{path_info}; }

sub forbidden_response {
    my $self = shift;
    $self->res->status(403);
    $self->res->body('forbidden');
}

sub notfound_response {
    my $self = shift;
    $self->res->status(404);
    $self->res->body('not found');
}

sub redirect_response {
    my ( $self, $url ) = @_;
    $self->res->redirect($url);
}

sub AUTOLOAD {
    my $self   = shift;
    my $caller = caller(0);
    ( my $method = $AUTOLOAD ) =~ s/.*:://;
    croak(" App can be extended only by Plugins [ !$caller->$method ] ")
      unless ( $caller->isa('Cot::Plugin') );
    no strict 'refs';
    *$method = sub {
        my $self = shift;
        $self->{$method} = $_[0] if ( $_[0] );
        return $self->{$method};
    };
    $self->$method(@_);
}

sub DESTROY { }

sub _export_run {
    my @argv   = @_;
    my $class  = caller(0);
    my $runner = Plack::Runner->new;
    if ( scalar @argv ) {
        @argv = split( /\s+/, $argv[0] ) if ( scalar @argv == 1 );
        $runner->parse_options(@argv);
    }
    my $app = sub { $class->app(shift); };
    $runner->run($app);
}

1;
__END__

=encoding utf-8

=head1 NAME

Cot - super lightweight perl framework based on Plack

=head1 SYNOPSIS

    #!/usr/bin/env perl -w

    use Cot;

    get '/' => sub {
        my $self = shift;
        $self->res->status(200);
        $self->res->headers( { 'Content-Type' => 'text/plain', } );
        $self->res->body('Hello world!');
    };

    run;

The above is a basic but functional web app created with Cot.

=head1 DESCRIPTION

Cot is super lightweight perl framework base on Plack!
you can extend Cot by using many plugins.



You can initialize application skelton using cotto utility executable.

    % cot init Test
    % cd ./Test
    % cot run

=head1 METHODS

=head2 plugin

Added plugins to Cot App.

    use Cot;

    plugin qw/Config Session/;

=head2 get

Receive GET request:

    use Cot;

    get '/api/echo' => sub {
        my $self = shift;
        #code
    };

=head2 post

Receive POST request:

    use Cot;

    post '/api/echo' => sub {
        my $self = shift;
        #code
    };

=head2 put

Receive PUT request:

    use Cot;

    put '/api/echo' => sub {
        my $self = shift;
        #code
    };

=head2 delete

Receive DELETE request:

    use Cot;

    delete '/api/echo' => sub {
        my $self = shift;
        #code
    };

=head2 options

Receive OPTIONS request:

    use Cot;

    options '/api/echo' => sub {
        my $self = shift;
        #code
    };

=head2 patch

Receive PATCH request:

    use Cot;

    patch '/api/echo' => sub {
        my $self = shift;
        #code
    };

=head2 any

Receive any HTTP request:

    use Cot;

    any '/api/echo' => sub {
        my $self = shift;
        #code
    };

=head2 static

Serve static files:

    use Cot;

    static => '/ui';

=head2 run

Up the Plack execution loop.
You can set L<plackup> arguments.

    use Cot;

    run("-- port 5001 -R public");

=head2 forbidden_response

    use Cot;

    get '/secret' => sub {
        $self->res->status(403);
        $self->res->body('forbidden');
    };

same as below.

    use Cot;

    get '/secret' => sub {
        $self->forbidden_response;
    };

=head2 notfound_response

    use Cot;

    get '/secret' => sub {
        $self->res->status(404);
        $self->res->body('notfound');
    };

same as below.

    use Cot;

    get '/secret' => sub {
        $self->notfound_response;
    };

=head2 redirect_response

    use Cot;

    get '/secret' => sub {
        $self->res->redirect('/', 301);
    };

same as below.

    use Cot;

    get '/secret' => sub {
        $self->redirect_response;
    };

=head1 Context METHODS

=head2 req

B<req> is a L<Plack::Request> Object. You can call all methods of Plack::Request.

    use Cot;

    get '/test' => sub {
        my $self = shift;
        my $req = $self->req;
        my $test = $req->param('test');
    };

=head2 res

B<res> is a L<Plack::Response> Object. You can call all methods of Plack::Response.

    use Cot;

    get '/test' => sub {
        my $self = shift;
        $self->res->status(200);
        $self->res->headers({'Content-Type' => 'text/plain' });
        $self->res->body($self->config->{sample});
    };

=head2 path_info

B<path_info> is PATH_INFO Array ref object.

    use Cot;

    # if called /test/hello
    get '/test' => sub {
        my $self = shift;
        my $info = $self->path_info->[0]; #hello
        ...
    };

=head2 env

B<env> is Plack environment variable.

    use Cot;

    get '/test' => sub {
        my $self = shift;
        my $remote_addr = $self->env->{REMOTE_ADDR}; # same as $self->req->address
        ...
    };

=head2 path

B<path> is requested PATH string

    use Cot;

    # if called /test/hello/myname
    get '/test' => sub {
        my $self = shift;
        my $path = $self->path; # /test/hello/myname
        ...
    };

=head1 ENV

You can set ENVIRONMENT variables for change behaviour.

=head2 COT_ROOT

Default value is " B <.>". For example mod_perl configuration, you can set

    PerlSetEnv COT_ROOT /www/TestApp/

=head2 COT_ENV

Default value is B<developement>. You can change COT_ENV for configration.

    #!/bin/sh
    export COT_ENV=production
    cot run

=head2 COT_DIRECTORYINDEX

Default value is none. If you use B<static> method, automatically serve DIRECTORYINDEX

   #!/bin/sh
   export DIRECTORYINDEX=index.html:index.xhtml
   cot run

=head1 PLUGINS

Cot has plaggable interface. For default install only L<Cot::Plugin::Config> can be used.

config file($ENV{COT_ROOT}/config.yaml):

    developement:
      hello: world

application code:

    use Cot;
    use Cot::Plugin qw/Config/;
    # or use Cot::Plugin::Config;

    get '/' => sub {
       my $self = shift;
       my $hello = $c->config->{'hello'}; # world
       ...
    };

=head1 AUTHORS

This module has been written by Yusuke Shibata <shibata@yusukeshibata.jp> and others,
see the AUTHORS file that comes with this distribution for details.

=head1 SOURCE CODE

The source code for this module is hosted on GitHub
L<https://github.com/yusukeshibata/Cot>.  Feel free to fork the repository and
submit pull requests!

=head1 DEPENDENCIES

The following modules are mandatory (Cot cannot run without them):

=over 8

=item L<YAML>

=item L<Plack>

=back

=head1 LICENSE

Copyright (C) Yusuke Shibata

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Yusuke Shibata E<lt>shibata@yusukeshibata.jpE<gt>
